### PR TITLE
Pass-through setattr for FabricModule

### LINKED ISSUE #17285: Implement `__setattr__` on `_FabricModule`
### Description & Motivation

We currently have only `__getattr__` implemented on the `_FabricModule`:

https://github.com/Lightning-AI/lightning/blob/c2d9ec4e1b9ab32b222630f3d094e657c65324f8/src/lightning/fabric/wrappers.py#L135-L143

Something like this works:

```py
model.x = 1
wrapped = fabric.setup(model)
print(wrapped.x)
```
But this doesn't:
```py
model.x = 1
wrapped = fabric.setup(model)
wrapped.x = 2
```
Current workaround is to do

```py
wrapped._original_module.x = 2
```

### Pitch

Implement  `__setattr__` on the `_FabricModule` and forward to the `_original_module`. This has to done carefully, as not to break the initialization of the `_FabricModule` and its attributes.

### Alternatives

_No response_

### Additional context

_No response_

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by bkiat1123**
## What does this PR do?

Implement __setattr__ on FabricModule.

Fixes #17285 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

File: src/lightning/fabric/wrappers.py [hunk 1]
```diff
@@ -179,6 +179,31 @@ def __getattr__(self, item: Any) -> Any:
             self._validate_method_access(item, attr)
             return attr
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        caller_func = inspect.stack()[1].function
+        if caller_func == "__init__":
+            super().__setattr__(name, value)
```
Comment (awaelchli) [inline src/lightning/fabric/wrappers.py:185]: This is a neat trick, but I'm wondering if it's save. In theory, and it is not so far fetched IMO, one could define a Trainer like this:

```py
class MyTrainer:
    def __init__(self, model, ...):
        self.fabric = Fabric(...)
        self.model = fabric.setup(model)
        self.model.trainer = self
        ...
        
```
I'm not saying the above is a good practice or anything. But in this case I believe we would see the Trainer's init in `inspect.stack()[1].function`, but the intention is to have this code only triggered for FabricModule's init.

File: tests/tests_fabric/test_wrappers.py [hunk 2]
```diff
@@ -104,6 +104,35 @@ def __init__(self, module):
     warning_cache.clear()
 
 
+def test_fabric_module_setattr():
```
Comment (awaelchli) [inline tests/tests_fabric/test_wrappers.py:107]: Could we extend this test case by adding 
1. A call that sets a submodule (simulating compliance with setting modules onto modules in PyTorch)
2. Defining a property-setter on the original module

These are some edge-cases that come to mind, so maybe good to assert they are working as expected

**Review by awaelchli — COMMENTED** [review]
Thanks for working on this. I left a couple comments, but tests pass so that's a good sign! 🎉

File: src/lightning/fabric/wrappers.py (hunk 1 above)
Comment (bkiat1123) [reply src/lightning/fabric/wrappers.py]: I agree it is not far fetched. I have resorted to defining a flag variable at class level, to help us keep track whether the instance is initialized.

File: tests/tests_fabric/test_wrappers.py [hunk 3]
```diff
@@ -104,6 +104,54 @@ def __init__(self, module):
...
+    fabric_module.new_attribute = 100
+    assert original_module.new_attribute == 100
+
+    # Modify existing attribute on original_module
+    fabric_module.attribute = 100
+    assert original_module.attribute == 100
+
+    # Check setattr of original_module
+    fabric_module.x = 100
+    assert original_module.x == 100
```
Comment (Borda) [inline tests/tests_fabric/test_wrappers.py:147]: lets rather use different numbers for each

**Review by awaelchli — APPROVED** [review]
Great work! I pushed a small cosmetic change and an additional two assertions.

File: pyproject.toml [hunk 4]
```diff
@@ -100,6 +100,7 @@ ignore-init-module-imports = true
     "S605",  # todo: Starting a process with a shell: seems safe, but may be changed in the future; consider rewriting without `shell`
     "S607",  # todo: Starting a process with a partial executable path
     "RET504",  # todo:Unnecessary variable assignment before `return` statement
+    "RET503",
```
Comment (awaelchli) [inline pyproject.toml:103]: Had to add this cc @Borda. Otherwise, pre-commit would force format the code with `return None` at the end.

Comment (Borda) [reply pyproject.toml:103]: Maybe raise the issue with Ruff? :thinking:

File: src/lightning/fabric/wrappers.py [hunk 5]
```diff
@@ -185,6 +189,28 @@ def __getattr__(self, item: Any) -> Any:
...
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        if not self._fabric_module_initialized:
+            return super().__setattr__(name, value)
+
+        # Get the _original_module attribute
+        original_module = self._original_module
+        original_has_attr = hasattr(original_module, name)
+        # Can't use super().__getattrr__ because nn.Module only checks _parameters, _buffers, and _modules
+        fabric_has_attr = name in self.__dict__
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:200]: Can't this be?

```suggestion
        fabric_has_attr = hasattr(self, name)
```

File: src/lightning/fabric/wrappers.py [hunk 6]
```diff
@@ -185,6 +189,28 @@ def __getattr__(self, item: Any) -> Any:
...
+
+        if not (original_has_attr or fabric_has_attr):
+            return setattr(original_module, name, value)
```
Comment (carmocca) [inline src/lightning/fabric/wrappers.py:203]: I might have missed previous discussions. Why doesn't this set it for both?

If they both define it, we set it for both
If neither of them define it, only in the original.

Why is that?

**Review by carmocca — COMMENTED** [review]
Some questions!

File: src/lightning/fabric/wrappers.py [hunk 7]
```diff
@@ -185,6 +189,28 @@ def __getattr__(self, item: Any) -> Any:
             self._validate_method_access(item, attr)
             return attr
 
+    def __setattr__(self, name: str, value: Any) -> None:
+        if not self._fabric_module_initialized:
+            return super().__setattr__(name, value)
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:194]: I did that to avoid the extra indentation: https://github.com/Lightning-AI/lightning/pull/17731/commits/61a1538cf531e1d89063bf25bfce2f131de36e87

File: src/lightning/fabric/wrappers.py (hunk 5 above)
Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:200]: because we have overridden getattr ourselves, and this line here was to avoid calling getattr.

File: src/lightning/fabric/wrappers.py (hunk 6 above)
Comment (awaelchli) [reply src/lightning/fabric/wrappers.py:203]: Because if the user does this

```
my_model.new_attribute = 1
```
Their intention is to set the attribute on the original module. It would be pointless to set a new attribute on our fabric wrapper and also on the original, because there is no mechanism to keep it in sync, and our getattr function would potentially return the wrong value.

**Review by carmocca — APPROVED** [review]
